Duplicated warning message in -Wcomma

|  |  |
| --- | --- |
| Bugzilla Link | [46438](https://llvm.org/bz46438) |
| Version | trunk |
| OS | Linux |
| CC | @zygoloid |

## Extended Description 
This code, bug.cc, clang-9 to clang-trunk emit two duplicated warning message "warning: possible misuse of comma operator here [-Wcomma]", while clang-6 to clang-8 just give one warning message about this.

$cat bug.cc
const long var = 10;
static_assert (1 ? ( 0 , var ) : 0 , "" );

$clang++-trunk -Wcomma -c bug.cc 
bug.cc:2:24: warning: possible misuse of comma operator here [-Wcomma]
static_assert (1 ? ( 0 , var ) : 0 , "" );
                       ^
bug.cc:2:22: note: cast expression to void to silence warning
static_assert (1 ? ( 0 , var ) : 0 , "" );
                     ^
                     static_cast<void>( )
bug.cc:2:24: warning: possible misuse of comma operator here [-Wcomma]
static_assert (1 ? ( 0 , var ) : 0 , "" );
                       ^
bug.cc:2:22: note: cast expression to void to silence warning
static_assert (1 ? ( 0 , var ) : 0 , "" );
                     ^
                     static_cast<void>( )
2 warnings generated.


$clang++-8 -Wcomma -c bug.cc 
bug.cc:2:24: warning: possible misuse of comma operator here [-Wcomma]
static_assert (1 ? ( 0 , var ) : 0 , "" );
                       ^
bug.cc:2:22: note: cast expression to void to silence warning
static_assert (1 ? ( 0 , var ) : 0 , "" );
                     ^
                     static_cast<void>( )
1 warning generated.

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -502,6 +502,9 @@ Bug Fixes in This Version
   The values of 0 and 1 block any unrolling of the loop. This keeps the same behavior with GCC.
   Fixes (`#88624 <https://github.com/llvm/llvm-project/issues/88624>`_).
 
+- Clang will no longer emit a duplicate -Wunused-value warning for an expression
+  `(A, B)` which evaluates to glvalue `B` that can be converted to non ODR-use. (#GH45783)
+
 Bug Fixes to Compiler Builtins
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 

--- a/clang/lib/Sema/SemaExpr.cpp
+++ b/clang/lib/Sema/SemaExpr.cpp
@@ -19707,18 +19707,17 @@ static ExprResult rebuildPotentialResultsAsNonOdrUsed(Sema &S, Expr *E,
       ExprResult Sub = Rebuild(LHS);
       if (!Sub.isUsable())
         return Sub;
-      LHS = Sub.get();
+      BO->setLHS(Sub.get());
     //   -- If e is a comma expression, ...
     } else if (BO->getOpcode() == BO_Comma) {
       ExprResult Sub = Rebuild(RHS);
       if (!Sub.isUsable())
         return Sub;
-      RHS = Sub.get();
+      BO->setRHS(Sub.get());
     } else {
       break;
     }
-    return S.BuildBinOp(nullptr, BO->getOperatorLoc(), BO->getOpcode(),
-                        LHS, RHS);
+    return ExprResult(BO);
   }
 
   //   -- If e has the form (e1)...

--- a/clang/test/CXX/drs/dr7xx.cpp
+++ b/clang/test/CXX/drs/dr7xx.cpp
@@ -28,10 +28,8 @@ namespace cwg712 { // cwg712: partial
         use(a);
         use((a));
         use(cond ? a : a);
-        // FIXME: should only warn once
         use((cond, a));
         // expected-warning@-1 {{left operand of comma operator has no effect}}
-        // expected-warning@-2 {{left operand of comma operator has no effect}}
 
         (void)a;
         // expected-error@-1 {{reference to local variable 'a' declared in enclosing function 'cwg712::f'}} FIXME

